### ISSUE TITLE
Consider dropping ShiftToggle from Jumbo rules

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4851]
I think the Jumbo rule set would be improved by not including the ShiftToggle rules in there, at all or mostly (maybe only include those for very short passwords). However, now that we had this in there, it's tricky to exclude them without restore of old sessions unintentionally skipping other rules. Ideas?

My testing of the Jumbo rules as currently in this repo, against NT hashes on one CPU core:

```
$ ./john -form=nt -verb=1 -encoding=cp1252 -w=pwned-passwords-1000plus-r -ru=jumbo pp-by-hash-10m
Loaded 10000000 password hashes with no different salts (NT [MD4 128/128 AVX 4x3])
Remaining 9766548 password hashes with no different salts
Press 'q' or Ctrl-C to abort, almost any other key for status
13158g 0:00:00:01 0.10% (ETA: 20:03:34) 12183g/s 2474Kp/s 2474Kc/s 26086GC/s Yvon..arn
51093g 0:00:00:02 0.20% (ETA: 20:03:35) 24563g/s 4505Kp/s 4505Kc/s 45643GC/s boysliep..bradley8p
79686g 0:00:00:03 0.27% (ETA: 20:05:22) 25872g/s 6193Kp/s 6193Kc/s 61812GC/s Harding10..Harley080
134628g 0:00:00:07 0.63% (ETA: 20:05:28) 19015g/s 6976Kp/s 6976Kc/s 68336GC/s ,rekord..,renae
149186g 0:00:00:09 0.86% (ETA: 20:04:25) 16448g/s 6822Kp/s 6822Kc/s 66504GC/s M.ama1976..M.amakitty
237891g 0:00:00:11 1.00% (ETA: 20:05:20) 21489g/s 7213Kp/s 7213Kc/s 70053GC/s Grace778..Graduate78
540604g 0:00:00:48 4.00% (ETA: 20:06:56) 11246g/s 8474Kp/s 8474Kc/s 79805GC/s Katty1226..Kayja
661024g 0:00:01:22 6.61% (ETA: 20:07:36) 8055g/s 8864Kp/s 8864Kc/s 82559GC/s mv246969..mv25011986
701968g 0:00:02:08 11.46% (ETA: 20:05:33) 5481g/s 7629Kp/s 7629Kc/s 70582GC/s 1q@w#E$r%t6Y7u8i..a1B2C#D4E5f^g7
701999g 0:00:09:02 85.47% (ETA: 19:57:31) 1295g/s 1823Kp/s 1823Kc/s 16856GC/s AsDFghJKLzxcVBnM..MaNCheSTErunITED
708435g 0:00:10:18 DONE (2021-10-22 19:57) 1144g/s 1617Kp/s 1617Kc/s 14974GC/s
```

Ditto with ShiftToggle removed:

```
$ ./john -form=nt -verb=1 -encoding=cp1252 -w=pwned-passwords-1000plus-r -ru=jumbo-mod pp-by-hash-10m
Loaded 10000000 password hashes with no different salts (NT [MD4 128/128 AVX 4x3])
Remaining 9766548 password hashes with no different salts
Press 'q' or Ctrl-C to abort, almost any other key for status
53577g 0:00:00:02 2.24% (ETA: 20:02:05) 26008g/s 4757Kp/s 4757Kc/s 47727GC/s friendssr..froggy1r
141463g 0:00:00:08 8.42% (ETA: 20:02:11) 17551g/s 7312Kp/s 7312Kc/s 71282GC/s goosebumPs..gryffindOr
599714g 0:00:01:07 63.81% (ETA: 20:02:21) 8944g/s 9244Kp/s 9244Kc/s 86361GC/s gracie99pt..granddadpt
688145g 0:00:01:24 78.21% (ETA: 20:02:23) 8187g/s 9313Kp/s 9313Kc/s 86585GC/s zy232629..zy236541
697156g 0:00:01:33 85.25% (ETA: 20:02:25) 7492g/s 9403Kp/s 9403Kc/s 87206GC/s r11011999q..r11021996q
706858g 0:00:01:39 DONE (2021-10-22 20:02) 7093g/s 9374Kp/s 9374Kc/s 87335GC/s
```

As you can see, the original gets stuck with almost no progress after ~10%, then achieves a little bit more progress closer to the end (once it's finally done ShiftToggle and reached the remaining included rule sets). Removing ShiftToggle cracks 99.777% of the same passwords in under 1/6th the time.

OTOH, the above is in wall clock time. For slow hashes, things wouldn't be nearly this bad: notice how the p/s rate reduced, suggesting the rules were skipping many input wordlist entries, which is what cost this much time. With ShiftToggle, we had `1617Kp/s*618s` = 999M candidates, without `9374Kp/s*99s` = 928M. So it's ~99.8% cracked in ~92.8% candidates. I think this still speaks against keeping (the full) ShiftToggle (maybe e.g. its first half for low lengths would be OK to keep).

Edit, documenting for completeness: the testing above is against a 10M sample of NT hashes from HIBP v7 with passwords that are in RockYou excluded (that's why "Remaining 9766548"). The wordlist is overlap of passwords that are in RockYou _and_ have a count of 1000+ near their hashes in HIBP v7. There are 200788 lines in that wordlist and it's 1673298 bytes (so average length of input words is below 8). Due to the exclusion of RockYou from the remaining hashes, the above testing ends up 100% out-of-sample (which is even a bit too much so, since in the real world there would in fact be some overlap in passwords between two independent samples). I think this detail isn't crucial to this specific test, though, and similar results would be seen with other sane tests as well.

Things become worse than the above with not-so-focused wordlists that tend to contain longer words. For example, the full RockYou has words longer by 1.4 chars on average. Per my math, that's roughly 2.6x more time spent on ShiftToggle (although can't really tell that from the average alone as the lengths distribution matters). This seems to match my experience.

**Comment by solardiz** [comment]
Further, we already have case toggle (not shift toggle, but similar) for lengths up to 8 in Extra, which is also included in Jumbo through Single-Extra. Given this finding, I think we should fully drop ShiftToggle from Jumbo (if it weren't for restore issues...)

We may also want to "upgrade" the rules in Extra from case to shift toggling. Since we try all combinations of positions (not) to toggle, I think shift toggling produces a strict superset of candidate passwords of what case toggling does, right? (To be confirmed by testing.)

**Comment by solardiz** [comment]
If we remove ShiftToggle from Jumbo, we should probably add it to somewhere in All, or possibly only its end that doesn't overlap with the (to be) upgraded rules in Extra.

**Comment by solardiz** [comment]
Similar tests with RockYou 3+ hits (1136144 words) as wordlist and all 10M hashes (no pre-exclusion).

Original Jumbo rules:

```
1412096g 0:01:04:52 DONE (2021-10-22 23:03) 362.7g/s 1576Kp/s 1576Kc/s 13988GC/s !abong..
```

with ShiftToggle removed:

```
1409177g 0:00:10:51 DONE (2021-10-22 22:09) 2161g/s 8340Kp/s 8340Kc/s 74280GC/s !re!!r..
```

`8340*651/(1576*(64*60+52))` = 88.5% candidates, 1409177/1412096 = 99.8% cracks.

**Comment by magnumripper** [comment]
IMHO we shouldn't worry too much about restore issues. Bleeding is bleeding - and once we make a release, users should be (made) aware that restore of a session using rules isn't 100% supported between releases and if that's a problem, old jobs should run to completion using the older version. How else would we handle it? Never amend or delete any existing rule ever?

I think we should upgrade the 'Extra' rules to shift-toggling, perhaps not (see below) drop the overlapping part from 'ShiftToggle' but move the latter from 'Jumbo' to 'All'.

BTW if we ensure the toggling rules (after pp and optimizer) in 'Extra' is an identical subset of 'ShiftToggle', our dupe rules suppression logic will take care of the overlapping part.

Another thing hit me now: We could possibly enhance the `WN` command for case insensitive hashes: I believe the current code will produce wads of logically identical guesses where just letters got case shifted. We should be able to make it only toggle non-letters in that situation.

**Comment by magnumripper** [comment]
> Another thing hit me now: We could possibly enhance the `WN` command for case insensitive hashes: I believe the current code will produce wads of logically identical guesses where just letters got case shifted. We should be able to make it only toggle non-letters in that situation.

Looking into this, I realized I already addressed it canonically in 8bfc5998cc1, with zero performance overhead. I'm a genius with poor memory!

**Comment by solardiz** [comment]
> I think we should upgrade the 'Extra' rules to shift-toggling, perhaps not (see below) drop the overlapping part from 'ShiftToggle' but move the latter from 'Jumbo' to 'All'.

I'm about to move `ShiftToggle` from `Jumbo` to `All`, but I don't currently have time to bother with testing an upgrade of `Extra` rules to shift-toggling - this currently feels like unnecessary scope creep for my rules rework and it has non-trivial benefits vs. drawbacks (e.g., the `-c` rule rejects on those rules in `Extra` will no longer be right if we shift-toggle).
